db/similarity.py

```diff
@@ -0,0 +1,133 @@
+            if not result.rowcount:
+                print("Metric {} added for all recordings".format(metric.name))
+                break
+
+            for row in result.fetchall():
+                submit_similarity_by_id(row["id"], metrics=metrics)
+
+            offset = count_similarity()
+            print("Processed {} / {} ({:.3f}%)".format(offset,
+                                                       lowlevel_count,
+                                                       float(offset) / lowlevel_count * 100))
+
+
+def insert_similarity(id, vectors_info):
+    """Inserts a row of similarity vectors for a given lowlevel.id into
+    the similarity table.
+
+        Args: lowlevel.id to be submitted
+              vectors_info, list of tuples of the form:
+              (metric_name, vector, isnan)
+    """
+    with db.engine.connect() as connection:
+        values = []
+        for metric, vector, isnan in vectors_info:
+            value = ('ARRAY' + ('[' + ', '.join(["'NaN'::double precision"] *
+                     len(vector)) + ']' if isnan else str(list(vector))))
+            values.append(value)
+
+        values_string = ', '.join(values)
+        query = text("""
+            INSERT INTO similarity (
+                        id,
+                        mfccs,
+                        mfccsw,
+                        gfccs,
+                        gfccsw,
+                        key,
+                        bpm,
+                        onsetrate,
+                        moods,
+                        instruments,
+                        dortmund,
+                        rosamerica,
+                        tzanetakis)
+                 VALUES (
+                        :id,
+                        %(values)s)
+            ON CONFLICT (id)
+             DO NOTHING
+        """ % {"values": values_string})
+        connection.execute(query, {'id': id})
+
+
+def count_similarity():
+    # Get total number of submissions in similarity table
+    with db.engine.connect() as connection:
+        query = text("""
+            SELECT COUNT(*)
+              FROM similarity
+        """)
+        result = connection.execute(query)
+        return result.fetchone()[0]
+
+
+def submit_similarity_by_id(id, metrics=None):
+    """Computes similarity metrics for a single recording specified
+    by lowlevel.id, then inserts the metrics as a new row in the
+    similarity table."""
+    try:
+        id = int(id)
+    except ValueError:
+        raise BadDataException('Parameter `id` must be an integer.')
+
+    # Check that lowlevel submission exists for given id
+    if not db.data.check_for_submission(id):
+        raise NoDataFoundException('No submission for parameter `id`.')
+
+    if not metrics:
+        metrics = similarity.utils.init_metrics()
+
+    vectors_info = []
+    for metric in metrics:
+        data = metric.get_data(id)
```

Comment:
I'm sorry that I didn't catch this part earlier - I only just saw it now that I'm running through the whole insertion process.
This step is really inefficient because we're calling get_data once for each metric, but using exactly the same id. This means that for each id in the database we're running:
* 12 queries to get data for each metric
* 1 query to validate the id
* 1 query to do the actual insertion

multiplied by 10 million rows, this is going to take a _really_ long time.
We should change this so that we get the lowlevel data in a single sql query, and then each metric acts on the returned json data and selects the field that we need. This means we can only do one query per id.
We could leave the get_data functionality in case we need it in the future. We can add another method to each metric.

A further improvement would be to select all lowlevel items for a batch in a single query (e.g. 1 query instead of 1000), but that could be done as a later improvement